### PR TITLE
IGNITE-20818 Optimize GC in MV storage

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-20818

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/PersistentPageMemoryDataRegionConfigurationSchema.java [hunk 1]
```diff
@@ -29,7 +30,8 @@
 @Config
 public class PersistentPageMemoryDataRegionConfigurationSchema extends BasePageMemoryDataRegionConfigurationSchema {
     /** Default size. */
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/PersistentPageMemoryDataRegionConfigurationSchema.java:32]: Thank you

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RemoveWriteOnGcInvokeClosure.java [hunk 2]
```diff
@@ -76,14 +79,16 @@ public void call(@Nullable VersionChain oldRow) throws IgniteInternalCheckedExce
 
         result = nextRowVersion;
 
+        assert result.nextLink() == NULL_LINK;
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RemoveWriteOnGcInvokeClosure.java:82]: Ok

**Review by tkalkirill — APPROVED** [review]

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/PersistentPageMemoryDataRegionConfigurationSchema.java [hunk 3]
```diff
@@ -28,8 +29,9 @@
  */
 @Config
 public class PersistentPageMemoryDataRegionConfigurationSchema extends BasePageMemoryDataRegionConfigurationSchema {
-    /** Default size. */
-    public static final long DFLT_DATA_REGION_SIZE = 256 * MiB;
+    /** Default size, maximum between 256 MB and 20% of the total physical memory. */
+    @SuppressWarnings("NumericCastThatLosesPrecision")
+    public static final long DFLT_DATA_REGION_SIZE = Math.max(256 * MiB, (long) (0.2 * getTotalMemoryAvailable()));
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/PersistentPageMemoryDataRegionConfigurationSchema.java:34]: We use Math.max. If info is not available, size will bee 256MiB. This is not fully reflected in the javadoc, I may fix that

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/LowWatermarkConfigurationSchema.java [hunk 4]
```diff
@@ -41,7 +41,7 @@ public class LowWatermarkConfigurationSchema {
     // TODO https://issues.apache.org/jira/browse/IGNITE-18977 Make these values configurable and create dynamic validator after that.
     @Range(min = DEFAULT_IDLE_SAFE_TIME_PROPAGATION_PERIOD_MILLISECONDS + CLOCK_SKEW)
     @Value(hasDefault = true)
-    public long dataAvailabilityTime = TimeUnit.MINUTES.toMillis(45);
+    public long dataAvailabilityTime = TimeUnit.MINUTES.toMillis(10);
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/LowWatermarkConfigurationSchema.java:44]: Everything's completely random. We will figure out our defaults later.
Using 10 instead of 45 allows you to run shorter tests to see the difference (with default settings)

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/PersistentPageMemoryDataRegionConfigurationSchema.java [hunk 5]
```diff
@@ -29,7 +29,10 @@
  */
 @Config
 public class PersistentPageMemoryDataRegionConfigurationSchema extends BasePageMemoryDataRegionConfigurationSchema {
-    /** Default size, maximum between 256 MB and 20% of the total physical memory. */
+    /**
+     * Default size, maximum between 256 MiB and 20% of the total physical memory.
+     * 256 MiB, is system was unable to retrieve physical memory size.
```
Comment (rpuch) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/PersistentPageMemoryDataRegionConfigurationSchema.java:34]:
```suggestion
     * 256 MiB, if system was unable to retrieve physical memory size.
```

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/VolatilePageMemoryDataRegionConfigurationSchema.java [hunk 6]
```diff
@@ -29,7 +29,10 @@
  */
 @Config
 public class VolatilePageMemoryDataRegionConfigurationSchema extends BasePageMemoryDataRegionConfigurationSchema {
-    /** Default initial size, maximum between 256 MB and 20% of the total physical memory. */
+    /**
+     * Default initial size, maximum between 256 MiB and 20% of the total physical memory.
+     * 256 MiB, is system was unable to retrieve physical memory size.
```
Comment (rpuch) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/configuration/schema/VolatilePageMemoryDataRegionConfigurationSchema.java:34]:
```suggestion
     * 256 MiB, if system was unable to retrieve physical memory size.
```

**Review by rpuch — APPROVED** [review]
